src/AspNet.Security.OAuth.Yammer/YammerAuthenticationHandler.cs

```diff
@@ -89,13 +93,47 @@ protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] st
             // Note: Yammer doesn't return a standard OAuth2 response. To make this middleware compatible
             // with the OAuth2 generic middleware, a compliant JSON payload is generated manually.
             // See https://developer.yammer.com/docs/oauth-2 for more information about this process.
-            var payload = JObject.Parse(await response.Content.ReadAsStringAsync())["access_token"].Value<JObject>();
-            payload["access_token"] = payload["token"];
-            payload["token_type"] = string.Empty;
-            payload["refresh_token"] = string.Empty;
-            payload["expires_in"] = string.Empty;
+            string accessToken;
+
+            using (var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync()))
+            {
+                accessToken = payload.RootElement.GetProperty("access_token").GetString("token");
+            }
+
+            // HACK Work out the best way to do this with System.Text.Json
+            using (var stream = new MemoryStream())
+            {
+                await WriteAccessTokenAsync(accessToken, stream);
+
+                var copy = JsonDocument.Parse(stream);
+                return OAuthTokenResponse.Success(copy);
+            }
+        }
+
+        private async Task WriteAccessTokenAsync(string token, Stream stream)
+        {
+            var output = new StreamPipeWriter(stream);
+
+            WriteAccessToken(token, output);
+
+            await output.FlushAsync();
+            output.Complete();
+
+            stream.Seek(0, SeekOrigin.Begin);
+        }
+
+        private void WriteAccessToken(string token, IBufferWriter<byte> bufferWriter)
+        {
+            var writer = new Utf8JsonWriter(bufferWriter);
+
+            writer.WriteStartObject();
+            writer.WriteString("access_token", token);
```

Comment:
@bartonjs is there an easier way to construct a JsonDocument from a fixed set of keys & values other than fully serializing and re-parsing it?

Reply:
> is there an easier way to construct a JsonDocument from a fixed set of keys & values other than fully serializing and re-parsing it?

Not currently, no.  It doesn't have the ability to create JSON, it just understands how to cursor over the text (semi-)intelligently.